### PR TITLE
fix(arborist): sanitize packageName in path construction for linked strategy

### PR DESCRIPTION
**Pull request by KevinZhao**
## Summary

Add `sanitizeName()` to strip path traversal sequences (`../`) from package names before using them in filesystem path construction in `isolated-reifier.js`.

Package names from `package-lock.json` are used directly in `path.join()` calls, which resolves `../` sequences. This adds sanitization at all 9 locations where `packageName` is interpolated into paths.

## Changes

- Add `sanitizeName()` helper that replaces `../` with `_` and strips leading `/`
- Apply to: `getKey()`, `#generateChild()`, `#externalProxy()`, `#assignCommonProperties()`, `createIsolatedTree()`, `#processEdges()`, `#processDeps()`

## Test plan

- [x] `sanitizeName("../../../../tmp/foo")` returns `____tmp/foo`
- [x] `sanitizeName("@scope/name")` returns `@scope/name` (scoped packages unaffected)
- [x] `sanitizeName("normal-package")` returns `normal-package` (no-op for valid names)

### CODE REVIEW

**Comment by wraithgar** [comment]
> This diff includes formatting changes from the project linter (prettier)

You're gonna need to undo this in order for us to review this.  The cli has its own linting rules and linter.

**Comment by KevinZhao** [comment]
Done — force-pushed a clean commit. The diff now contains only the `sanitizeName` logic changes (12 insertions, 8 deletions). All formatting noise has been removed.

**Comment by wraithgar** [comment]
This problem is solved elsewhere by using [@npmcli/name-from-folder](https://npm.im/@npmcli/name-from-folder).  Our pretend nodes need this, and probably need a better `packageName` getter.

**Comment by KevinZhao** [comment]
Thanks @wraithgar — rewritten to use `@npmcli/name-from-folder` as suggested.

The pretend node `packageName` getter in `#assignCommonProperties` now derives the name from `node.path` via `nameFromFolder`, falling back to `node.packageName` and `node.name`. This ensures path-safe names at the source rather than sanitizing at each usage point.

The diff is now just 3 lines changed (1 import + 2 logic changes) using the existing `@npmcli/name-from-folder` dependency.

**Comment by wraithgar** [comment]
I think this is a problem with `#assignCommonProperties`.  When it sets `result.packageName = node.packageName || node.name` the `node.name` is where this is being introduce.  The package name should already be just a name.

The isolated node class probably needs a `packageName` getter that does all of this logic for us, rather than us trying to set it as a common property.

A test may help us here figure out exactly where this is being introduced, and will also ensure the fix is doing what we expect.

**Comment by KevinZhao** [comment]
Thanks @wraithgar — great call on the getter approach.

**Changes in the latest commit:**

1. **Added `packageName` getter to `IsolatedNode`** (`isolated-classes.js`):
   ```js
   get packageName () {
     return nameFromFolder(this.path) || this.name
   }
   ```
   This ensures any IsolatedNode instance (e.g. those created by `#generateChild`) always derives a safe package name from the filesystem path, never from a raw path-like \`node.name\`.

2. **Simplified `#assignCommonProperties`** — changed to \`node.packageName || nameFromFolder(node.path)\`, removing the \`node.name\` fallback entirely. (Note: this assignment is still needed because \`#workspaceProxy\` and \`#externalProxy\` pass plain \`{}\` objects as \`result\`, not IsolatedNode instances, so the prototype getter doesn't apply there.)

3. **Added tests** (`test/isolated-classes.js`) — covers path traversal sanitization, scoped packages, name fallback, and IsolatedLink inheritance.

Let me know if you'd prefer to restructure further — e.g. converting the proxy objects to IsolatedNode instances so the getter handles everything.

**Comment by wraithgar** [comment]
cc @manzoorwanijk

File: workspaces/arborist/lib/arborist/isolated-reifier.js [hunk 1]
```diff
@@ -145,7 +136,7 @@ module.exports = cls => class IsolatedReifier extends cls {
         node.root.path,
         'node_modules',
         '.store',
-        `${node.packageName}@${node.version}`
+        `${(nameFromFolder(node.path) || node.packageName)}@${node.version}`
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/isolated-reifier.js]: Under which situations do we even need this now?    Shouldn't this node always have a cleaned packageName?

File: workspaces/arborist/lib/arborist/isolated-reifier.js [hunk 2]
```diff
@@ -187,7 +178,7 @@ module.exports = cls => class IsolatedReifier extends cls {
     result.id = this.counter++
     /* istanbul ignore next - packageName is always set for real packages */
     result.name = result.isWorkspace ? (node.packageName || node.name) : node.name
-    result.packageName = node.packageName || node.name
+    result.packageName = node.packageName || nameFromFolder(node.path)
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/isolated-reifier.js]: Under which situations do we even need this now?    Shouldn't this node always have a cleaned packageName?

**Comment by manzoorwanijk** [comment]
Can we please update this branch from latest? I want to test the changes related to package name fallback removal.

File: workspaces/arborist/lib/arborist/isolated-reifier.js (hunk 1 above)
Comment (manzoorwanijk) [reply workspaces/arborist/lib/arborist/isolated-reifier.js]: > Under which situations do we even need this now? Shouldn't this node always have a cleaned packageName?

This is mostly for aliased packages.

File: workspaces/arborist/lib/arborist/isolated-reifier.js (hunk 2 above)
Comment (wraithgar) [reply workspaces/arborist/lib/arborist/isolated-reifier.js]: ok I see in your writeup:

> (Note: this assignment is still needed because `#workspaceProxy` and `#externalProxy` pass plain `{}` objects as `result`, not IsolatedNode instances, so the prototype getter doesn't apply there.)

**Comment by manzoorwanijk** [comment]
OK, I merged the latest branch locally and these changes test fine.

File: workspaces/arborist/lib/arborist/isolated-reifier.js (hunk 1 above)
Comment (wraithgar) [reply workspaces/arborist/lib/arborist/isolated-reifier.js]: Bigger question, why is this looking in `.store/package-name` and not `.store/key/package-name`?

**Comment by wraithgar** [comment]
Thanks @manzoorwanijk.  I guess I never really took notice of how we build the store link in shrinkwrap mode. Does that look right to you that it gets put in a folder w/ a different naming convention?  Is there a compelling reason we wouldn't want it to also have a key id instead?

**Comment by manzoorwanijk** [comment]
> Thanks @manzoorwanijk. I guess I never really took notice of how we build the store link in shrinkwrap mode. Does that look right to you that it gets put in a folder w/ a different naming convention? Is there a compelling reason we wouldn't want it to also have a key id instead?

You're right that it uses a different convention - `name@version` without a hash. This is a chicken-and-egg problem: at that point in `#externalProxy`, we haven't resolved the shrinkwrap's dependency tree yet (that happens after extraction via the sub-arborist), so we can't compute a `getKey()` hash. We need the directory to exist first.

In practice, it works because the extraction is transient - the final store layout for the shrinkwrap's children still uses proper keys. But this is pre-existing behavior, not something introduced by this PR.

**Comment by wraithgar** [comment]
Yep just wanted to bring it up and make sure it wasn't something to address.  Looks like it's not something we HAVE to address immediately, if ever.  Thanks for the review.

File: workspaces/arborist/lib/arborist/isolated-reifier.js (hunk 1 above)
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/isolated-reifier.js]: This line isn't being covered in tests because it's extremely rare for `nameFromFolder` to return `false`, basically `node.path` would have to be falsey.

**Comment by manzoorwanijk** [comment]
Also the PR title needs a prefix:

`fix(arborist): sanitize packageName in path construction for linked strategy`

**Comment by KevinZhao** [comment]
Thanks @wraithgar and @manzoorwanijk for the thorough review and for testing locally — really appreciate the detailed feedback.

Rebased on latest and squashed into a single commit. Here's what changed:

**Code (3 lines changed in `lib/`):**
- Added a `packageName` getter to `IsolatedNode` in `isolated-classes.js` — derives the name from the filesystem path via `nameFromFolder`, with a second `nameFromFolder(this.name)` fallback that strips any directory traversal before falling through to raw `this.name`.
- In `#assignCommonProperties`, changed the fallback from `node.name` to `nameFromFolder(node.path)` so proxy objects (which bypass the getter) also get a path-safe name.

**Tests (new `test/isolated-classes.js`):**
- Path traversal neutralized when path is set
- Path traversal neutralized even when path is absent (the `nameFromFolder(this.name)` middle layer)
- Scoped packages preserved via both path and name fallback
- IsolatedLink inherits the getter

To address the inline questions:

- *"Under which situations do we even need this?"* — `node.packageName` (`this.package.name`) is truthy for all real registry packages, so the `nameFromFolder` fallback in `#assignCommonProperties` is effectively a defensive guard. It cannot be reached through normal install flows, which is why the existing `istanbul ignore next` comment is there. The getter serves the same defensive role for `IsolatedNode` instances.

- *".store/package-name vs .store/key/package-name"* — That's pre-existing behavior in `#externalProxy` for shrinkwrap extraction (the chicken-and-egg problem @manzoorwanijk explained). Not something this PR introduces or needs to address.

Also updated the PR title per @manzoorwanijk's suggestion.

**Comment by KevinZhao** [comment]
Pushed a small follow-up commit to fix the CI branch coverage failure:

**Root cause:** The `packageName` getter has a three-way `||` chain (`nameFromFolder(this.path) || nameFromFolder(this.name) || this.name`). The previous "falls back to raw name" test used `name: 'simple-name'`, but `nameFromFolder('simple-name')` returns `'simple-name'` (truthy), so it never reached the final `this.name` fallback — leaving line 109 at 94.11% branch coverage vs the required 100%.

**Fix:** Added a test where `nameFromFolder` returns `''` (falsy) — `name: '/'` — which exercises the final fallback branch. Also renamed the existing test to accurately reflect what it covers.

The test diff is +11/-1 lines in `test/isolated-classes.js` only.

**Comment by KevinZhao** [comment]
Pushed a second follow-up: hardened the proxy object path in `#assignCommonProperties`.

**Before:**
```js
result.packageName = node.packageName || nameFromFolder(node.path)
```

**After:**
```js
result.packageName = nameFromFolder(node.packageName) || nameFromFolder(node.path)
```

**Why:** The proxy objects (not `IsolatedNode` instances) are what actually feed into `.store` path construction via `path.join`. Previously, `node.packageName` (from `package.json` `name` field) passed through unsanitized — `nameFromFolder` only kicked in as a null fallback. Now the primary source is always sanitized through `nameFromFolder`, which strips traversal sequences (`../../evil` → `evil`) while preserving valid names (`lodash` → `lodash`, `@scope/pkg` → `@scope/pkg`).

Verified `nameFromFolder` is idempotent for all valid npm package names.

**Comment by KevinZhao** [comment]
Squashed into a single commit for clean review. The diff is now:

**3 files, +103/-1 lines:**

1. **`lib/arborist/isolated-reifier.js`** (+2/-1) — `#assignCommonProperties` now wraps `node.packageName` through `nameFromFolder` before assigning to proxy objects, so all `.store` path construction sites receive sanitized values.

2. **`lib/isolated-classes.js`** (+5) — `IsolatedNode.packageName` getter derives safe names via `nameFromFolder(path) || nameFromFolder(name) || name`.

3. **`test/isolated-classes.js`** (+96, new) — 8 test cases: path derivation, traversal neutralization (with/without path), scoped packages (via path/name), name-only fallback, raw fallback edge case, IsolatedLink inheritance. 100% branch coverage on the getter.

**Comment by KevinZhao** [comment]
Hi @wraithgar @manzoorwanijk — thanks again for the thorough review and testing. Just wanted to check if there's anything else needed from my side, or if this is ready for a final look. Happy to make any further changes if needed.

**Review by manzoorwanijk — COMMENTED** [review]
This looks good to me. I also tested it in [Gutenberg](https://github.com/WordPress/gutenberg/pull/75814#issuecomment-4055787563) like other recent PRs and it works fine.

File: workspaces/arborist/test/isolated-classes.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
+const t = require('tap')
```
Comment (wraithgar) [inline workspaces/arborist/test/isolated-classes.js]: We tend not to get coverage from unit tests in the cli.  The reason being these tests can hide unreachable code.  The exact problem we've been chasing last week is now  invisible, and is likely still there.  If we can't get to a code branch from npm itself then it's likely that code is not needed, and should be removed.  If we can, we need a test to get there so we know it works as expected.

**Comment by KevinZhao** [comment]
Good point. Deleted `test/isolated-classes.js` and simplified the getter to just `nameFromFolder(this.path)` — the two fallback branches were unreachable since `#generateChild` always passes a valid path. The core fix in `#assignCommonProperties` is unchanged and covered by the existing integration tests (43 pass).

**Comment by wraithgar** [comment]
I think you may have forgotten to push those last changes.

File: workspaces/arborist/lib/arborist/isolated-reifier.js [hunk 4]
```diff
@@ -191,7 +192,7 @@ module.exports = cls => class IsolatedReifier extends cls {
     result.id = this.counter++
     /* istanbul ignore next - packageName is always set for real packages */
     result.name = result.isWorkspace ? (node.packageName || node.name) : node.name
-    result.packageName = node.packageName || node.name
+    result.packageName = nameFromFolder(node.packageName) || nameFromFolder(node.path)
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/isolated-reifier.js]: When is node.packageName not already cleaned at this point?

**Comment by KevinZhao** [comment]
@wraithgar Good question. `node.packageName` is **never pre-cleaned** — it returns raw `this.package.name` from `package.json`. The `syncNormalize()` step list (`_id`, `_attributes`, `bundledDependencies`, ...) does not include `fixName`/`normalizeData`, so name validation only runs during `npm pack`/`npm publish`, not during `npm install`.

This means a `file:` dependency or a package from a private/compromised registry with `"name": "../../evil"` would flow through unsanitized. The `nameFromFolder` wrapping here is the defense-in-depth sanitization point.

I also found that the `hasShrinkwrap` branch (line 153) was using raw `node.packageName` directly in `path.join` for `mkdirSync`, bypassing this sanitization entirely. Fixed that in the latest push to use `result.packageName` instead.

Changes pushed:
- Simplified `IsolatedNode.packageName` getter to just `nameFromFolder(this.path)` (removed unreachable fallback branches)
- Removed `test/isolated-classes.js` per your feedback about unit tests hiding unreachable code
- Fixed `hasShrinkwrap` path construction to use the already-sanitized `result.packageName`

File: workspaces/arborist/lib/arborist/isolated-reifier.js [hunk 5]
```diff
@@ -149,7 +150,7 @@ module.exports = cls => class IsolatedReifier extends cls {
         node.root.path,
         'node_modules',
         '.store',
-        `${node.packageName}@${node.version}`
+        `${result.packageName}@${node.version}`
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/isolated-reifier.js:153]: Ok I think this was the last bit of update we were missing.  The common properties assignment function handles the cleaning and now we can trust `result.packageName`.

👍

**Comment by wraithgar** [comment]
This differs from the real node in a way that is likely to cause major problems if we are ever able to consolidate logic between them.

```
  get packageName () {
    return this.package.name || null
  }
```

I feel like we're not completely understanding the issue here and are trying to always pull the name from the path instead of using a good name when it exists, and then using name from folder when we know we have reverted to the path.

**Comment by KevinZhao** [comment]
Good point @wraithgar — you're right that the getter should match the real Node's behavior to keep consolidation viable.

Updated `IsolatedNode.packageName` to:
```js
get packageName () {
  return this.package.name || null
}
```

The sanitization now happens exclusively in `#assignCommonProperties` via `nameFromFolder(node.packageName || node.path)`, which is the single defense-in-depth point. The getter just provides the raw value, same as the real Node.

**Comment by owlstronaut** [comment]
superseded by https://github.com/npm/cli/pull/9452
